operator/controllers/kyma_controller.go

```diff
@@ -504,3 +522,71 @@ func (r *KymaReconciler) GetModules(ctx context.Context, kyma *v1alpha1.Kyma) (p
 
 	return modules, nil
 }
+
+func (r *KymaReconciler) UpdateAndRemoveNotExistsModuleInfo(ctx context.Context, kyma *v1alpha1.Kyma) bool {
+	requireUpdateCondition := false
+	moduleInfoMap := kyma.GetModuleInfoMap()
+	moduleInfos := kyma.GetNotExistsModuleInfos()
+	if len(moduleInfos) == 0 {
+		return false
+	}
+	for i := range moduleInfos {
+		moduleInfo := moduleInfos[i]
+		_, err := r.getModule(ctx, moduleInfo)
+		if k8serrors.IsNotFound(err) {
+			requireUpdateCondition = true
+			delete(moduleInfoMap, moduleInfo.ModuleName)
+		}
+	}
+	kyma.Status.ModuleInfos = convertToNewModuleInfos(moduleInfoMap)
+	return requireUpdateCondition
+}
+
+func (r *KymaReconciler) DeleteNotExistsModules(ctx context.Context, moduleInfos []*v1alpha1.ModuleInfo) error {
+	var err error
+	if len(moduleInfos) == 0 {
+		return nil
+	}
+	for i := range moduleInfos {
+		moduleInfo := moduleInfos[i]
+		err = r.deleteModule(ctx, moduleInfo)
+	}
+
+	if client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("error deleting module %w", err)
+	}
+	return nil
+}
+
+func convertToNewModuleInfos(moduleInfoMap map[string]*v1alpha1.ModuleInfo) []v1alpha1.ModuleInfo {
+	newModuleInfos := make([]v1alpha1.ModuleInfo, 0)
+	for _, moduleInfo := range moduleInfoMap {
+		newModuleInfos = append(newModuleInfos, *moduleInfo)
+	}
+	return newModuleInfos
+}
+
+func (r *KymaReconciler) deleteModule(ctx context.Context, moduleInfo *v1alpha1.ModuleInfo) error {
+	module := unstructured.Unstructured{}
+	module.SetNamespace(moduleInfo.Namespace)
+	module.SetName(moduleInfo.Name)
+	module.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   moduleInfo.GroupVersionKind.Group,
+		Version: moduleInfo.GroupVersionKind.Version,
+		Kind:    moduleInfo.GroupVersionKind.Kind,
+	})
+	return r.Delete(ctx, &module, &client.DeleteOptions{})
+}
+
+func (r *KymaReconciler) getModule(ctx context.Context,
```

Comment:
This logic is used heavily also for the create/update check, maybe refactor to use this method there?

Reply:
indeed, replaced in two places, but also related to performance concern, we should not use it for get all modules (and unfortunately current are all for this purpose)